File: Easy/0069-sqrtx.ts
function mySqrt(x: number): number {
    if (x < 2) return x;
    
    let left = 2;
    // Use Math.floor to ensure integer division
    let right = Math.floor(x / 2); 
    let ans = 1;
    
    while (left <= right) {
        // Use Math.floor to ensure mid remains a whole number
        let mid = left + Math.floor((right - left) / 2);
        
        // Math.floor(x / mid) replicates C++ integer division behavior
        if (mid <= Math.floor(x / mid)) {
            ans = mid;       
            left = mid + 1;  
        } else {
            right = mid - 1; 
        }
    }
    return ans;
}
/*
More efficient solution for .TS
function mySqrt(x: number): number {
    let r = x;
    
    // Iteratively approach the true square root
    while (r * r > x) {
        r = Math.floor((r + x / r) / 2);
    }
    
    return r;
}
*/ 

File: Easy/0069-sqrtx.java
class Solution {
    public int mySqrt(int x) {
        if (x < 2) return x;
    
        int left = 2;
        int right = x / 2;
        int ans = 1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            // Using division instead of (mid * mid <= x) to prevent integer overflow
            if (mid <= x / mid) {
                ans = mid;       // mid could be the answer, save it
                left = mid + 1;  // Look for a larger valid integer
            } else {
                right = mid - 1; // mid is too large, look in the lower half
            }
        }
        return ans;
    }
}
/*
Newton's Method rather then binary search
class Solution {
    public int mySqrt(int x) {
        long r = x;
        
        // Iteratively average the guess until r * r is no longer greater than x
        while (r * r > x) {
            r = (r + x / r) / 2;
        }
        
        return (int) r;
    }
}
 */
